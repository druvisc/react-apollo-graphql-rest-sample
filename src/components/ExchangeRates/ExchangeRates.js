import React from 'react'
import gql from 'graphql-tag'
import { useLazyQuery } from '@apollo/react-hooks'

const GET_EXCHANGE_RATES = gql`
  query exchangeRatesQuery {
    exchangeRates @rest(type: "ExchangeRates", endpoint: "exchangerates" path: "latest/") {
      date
      base
      rates
    }
  }
`

function ExchangeRates() {
  const [
    getExchangeRates, { called, loading, error, data, refetch }
  ] = useLazyQuery(GET_EXCHANGE_RATES, {
    notifyOnNetworkStatusChange: true
  })

  const Button = () =>
    <button onClick={() => called ? refetch() : getExchangeRates()}>
      Load exchange rates
    </button>

  const Rates = () => {
    if (!called) return null
    if (loading) return <p>Loading...</p>
    if (error) return <p>Couldn't load exchange rates</p>

    return (
      <table>
        <tbody>
          {Object.keys(data.exchangeRates.rates).map(k =>
            <tr key={k}>
              <td>{k}</td>
              <td>{data.exchangeRates.rates[k]}</td>
            </tr>)}
        </tbody>
      </table>
    )
  }

  return (
    <>
      <Button />
      <Rates />
    </>
  )
}

export default ExchangeRates

