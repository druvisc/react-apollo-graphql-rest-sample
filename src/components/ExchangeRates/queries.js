import { gql } from '@apollo/client'

export const GET_EXCHANGE_RATES_LATEST = gql`
  query ExchangeRatesLatestQuery {
    exchangeRates @rest(type: "ExchangeRates", endpoint: "exchangerates" path: "latest/") {
      date
      base
      rates
    }
  }
`

export const GET_EXCHANGE_RATES_DATE = gql`
  query ExchangeRatesQuery($date: String!) {
    exchangeRates(date: $date) @rest(type: "ExchangeRates", endpoint: "exchangerates" path: "{args.date}") {
      _id: date
      date
      base
      rates
    }
  }
`
