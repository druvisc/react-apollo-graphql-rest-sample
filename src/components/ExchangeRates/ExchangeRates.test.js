import React from 'react'
import { mount, render } from 'enzyme'
// import { render } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import ExchangeRates, { GET_EXCHANGE_RATES } from './ExchangeRates'

const mocks = [
  {
    request: {
      query: GET_EXCHANGE_RATES,
    },
    result: {
      data: {
        rates: {
          HKD: 8.6836,
        },
        base: 'EUR',
        date: '2020-01-07'
      }
    }
  },
  {
    request: {
      query: GET_EXCHANGE_RATES,
    },
    error: new Error('Oops')
  }
]

describe('<ExchangeRates />', () => {
  it('renders without crashing', () => {
    mount(<ExchangeRates />)
  })

  it('renders only button', () => {
    const wrapper = render(<ExchangeRates />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders loading state', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ExchangeRates />
      </MockedProvider>
    )

    wrapper.find('button').simulate('click')
    // expect(wrapper).toMatchSnapshot()
  })
})
