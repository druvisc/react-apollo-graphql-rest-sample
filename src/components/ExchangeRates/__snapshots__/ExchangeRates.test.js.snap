// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExchangeRates /> renders only button 1`] = `
<button>
  Load exchange rates
</button>
`;
